packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -173,34 +173,59 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
         if (this.autoSave === 'on') {
             this.toDisposeOnAutoSave.dispose();
             const handle = window.setTimeout(() => {
-                this.doSave(TextDocumentSaveReason.AfterDelay);
+                this.scheduleSave(TextDocumentSaveReason.AfterDelay);
             }, this.autoSaveDelay);
             this.toDisposeOnAutoSave.push(Disposable.create(() =>
                 window.clearTimeout(handle))
             );
         }
     }
 
-    protected async doSave(reason: TextDocumentSaveReason): Promise<void> {
-        if (this.resource.saveContents) {
-            await this.fireWillSaveModel(reason);
-            if (this.dirty) {
-                const content = this.model.getValue();
-                await this.resource.saveContents(content);
-                this.setDirty(false);
+    protected saved: Promise<void> = Promise.resolve();
+    protected saveCancellationTokenSource = new CancellationTokenSource();
+    protected scheduleSave(reason: TextDocumentSaveReason): Promise<void> {
+        this.saveCancellationTokenSource.cancel();
+        this.saveCancellationTokenSource = new CancellationTokenSource();
+        const token = this.saveCancellationTokenSource.token;
+        return this.saved = this.saved.then(() => this.doSave(reason, token));
```

Comment:
from gitter by @hexa00:

> reading this it seems that a scheduleSave will cancel the previous one if it is running... ? could it make it so that continously typing autosave will not be called.. I'm not sure if it shouln't be the other way around... finish the one that was there and try to schedule one after that?
it may be fine too the moment one stops typing for 500ms it will save...